polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -152,105 +153,109 @@ async fn run<Context>(
 	)
 	.await;
 	ctx.spawn_blocking("pvf-validation-host", task.boxed())?;
-
-	loop {
-		match ctx.recv().await? {
-			FromOrchestra::Signal(OverseerSignal::ActiveLeaves(_)) => {},
-			FromOrchestra::Signal(OverseerSignal::BlockFinalized(..)) => {},
-			FromOrchestra::Signal(OverseerSignal::Conclude) => return Ok(()),
-			FromOrchestra::Communication { msg } => match msg {
-				CandidateValidationMessage::ValidateFromChainState(
-					candidate_receipt,
-					pov,
-					executor_params,
-					timeout,
-					response_sender,
-				) => {
-					let bg = {
-						let mut sender = ctx.sender().clone();
-						let metrics = metrics.clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_chain_state();
-							let res = validate_from_chain_state(
-								&mut sender,
-								validation_host,
-								candidate_receipt,
-								pov,
-								executor_params,
-								timeout,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-chain-state", bg.boxed())?;
-				},
-				CandidateValidationMessage::ValidateFromExhaustive(
-					persisted_validation_data,
-					validation_code,
-					candidate_receipt,
-					pov,
-					executor_params,
-					timeout,
-					response_sender,
-				) => {
-					let bg = {
-						let metrics = metrics.clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_exhaustive();
-							let res = validate_candidate_exhaustive(
-								validation_host,
-								persisted_validation_data,
-								validation_code,
-								candidate_receipt,
-								pov,
-								executor_params,
-								timeout,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-exhaustive", bg.boxed())?;
-				},
-				CandidateValidationMessage::PreCheck(
-					relay_parent,
-					validation_code_hash,
-					response_sender,
-				) => {
-					let bg = {
-						let mut sender = ctx.sender().clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let precheck_result = precheck_pvf(
-								&mut sender,
-								validation_host,
-								relay_parent,
-								validation_code_hash,
-							)
-							.await;
-
-							let _ = response_sender.send(precheck_result);
-						}
-					};
-
-					ctx.spawn("candidate-validation-pre-check", bg.boxed())?;
-				},
+	let mut res = Ok(());
+	let sender = ctx.sender().to_owned();
+
+	let read_stream = stream::poll_fn(|c| loop {
```

Comment:
Another possible way for implementing this would be to use a async_semaphore.
```
let s = Semaphore::new(MAX_CONCURENT_REQUESTS))
```

s.acquire() before all the spawn and you move the resulting guard in the future.